draft-ietf-quic-recovery.md

```diff
@@ -518,14 +518,20 @@ or when Initial or Handshake keys are discarded. This ensures the PTO is always
 set based on the latest RTT information and for the last sent packet in the
 correct packet number space.
 
+When setting the PTO timer, the ApplicationData packet number space (Section
+4.1.1 of {{QUIC-TLS}}) MUST be ignored until the handshake completes. PTO is not
+armed because issues other than packet loss might prevent an endpoint from
+receiving an acknowledgement. A receiver might lack the keys necessary to
+process packets, or the sender might lack the keys necessary to process
+acknowledgments. This occurs when the client sends 0-RTT packets as the server
+might not accept 0-RTT and the client will not have 1-RTT keys for processing
+acknowledgments until the handshake completes. This occurs for 1-RTT packets as
+either endpoint might not have the ability to read until the handshake is
+complete.
+
 When ack-eliciting packets in multiple packet number spaces are in flight,
-the timer MUST be set for the packet number space with the earliest timeout,
-with one exception. The ApplicationData packet number space (Section 4.1.1
-of {{QUIC-TLS}}) MUST be ignored until the handshake completes. Not arming
-the PTO for ApplicationData prevents a client from retransmitting a 0-RTT
-packet on a PTO expiration before confirming that the server is able to
-decrypt 0-RTT packets, and prevents a server from sending a 1-RTT packet on
-a PTO expiration before it has the keys to process an acknowledgement.
+the timer MUST be set to the earlier value of the Initial and Handshake packet
+number spaces.
```

Comment:
I think this is a behavior change and I don't think it's always correct, since it causes the timer to never be armed for ApplicationData anytime anything in Handshake is in flight.  There are cases the Handshake keys have been dropped by the server and the Handshake data will never be acknowledged.

Reply:
Regarding if the new approach works, it does, because the server would make sure that all the handshake messages up to ServerFinished would be delivered to the client, and then the client would make sure that ClientFinished would be delivered to the server. At that moment, the server would be allowed to send 1-RTT probes.

Regarding if this is a design change rather than editorial, the answer depends on how we interpret the existing text. [Section 6.2.1](https://quicwg.org/base-drafts/draft-ietf-quic-recovery.html#section-6.2.1-6) currently states:
> When ack-eliciting packets in multiple packet number spaces are in flight, the timer MUST be set for the packet number space with the earliest timeout, with one exception. The ApplicationData packet number space (Section 4.1.1 of QUIC-TLS) MUST be ignored until the handshake completes.

There are two ways to interpret these two sentences, depending on if we consider the MUST in the second sentence to be a sub-condition of the first sentence, or consider the two sentences as listing two distinct requirements. @marten-seemann points out that the latter is what the pseudo-code does, and assuming that is the case, there is no design change.

Reply:
@ianswett When you drop keys, you're also supposed to drop the corresponding packets in that packet number space, so the Handshake packets in your example would not be considered outstanding.

That being said, I have to admit I don't understand why the existing text is ambiguous. @kazuho's change here (the one to the -recovery document), if I understand it correctly, is purely editorial.

Reply:
@ianswett This isn't behavior change. I've suggested a rephrase above that ought to make this clearer.

Reply:
To be clear, the confirmed/complete aspect is a behavior change, which means for example that the client only arms PTO for a single PN space at once.

But this updated text looks good, assuming we go with that design change.